### PR TITLE
[WIP] plugin/k8s_external implement transfer zone for k8s_external

### PR DESCRIPTION
**Pull request by fturib**
### 1. Why is this pull request needed and what does it do?
I need to be able to serve a secondary server from the Authoritative server of the public side of the Cluster.
In other word, I need AXFR to be implemented in `k8s_external` plugin

### 2. Which issues (if any) are related?
#2526 

### 3. Which documentation changes (if any) need to be made?
oops .. will update on next commit

WIP, still todo:
- Update the README with 'tranferTo' and the support for AXFR/IXFR
- Implement a smart SERIAL for k8s_external
- add UT for Transfer support

### CODE REVIEW

File: plugin/k8s_external/external.go [hunk 1]
```diff
@@ -90,11 +91,11 @@ func (e *External) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Ms
 
 	switch state.QType() {
 	case dns.TypeA:
-		m.Answer = e.a(svc, state)
+		m.Answer = e.a(svc, &state)
```
Comment (miekg) [inline plugin/k8s_external/external.go]: state is not a pointer anywhere in coredns, why here?

File: plugin/k8s_external/xfr.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// Transfer implements the Transferer interface.
+func (e *External) Transfer(ctx context.Context, state request.Request) (int, error) {
+
+	if !e.transferAllowed(state) {
+		return dns.RcodeRefused, nil
+	}
+
+	serviceList, _ := e.externalFunc(state)
+	records, _ := e.srv(serviceList, nil)
```
Comment (miekg) [inline plugin/k8s_external/xfr.go]: can we just make a new function that returns the all the records we need?

File: plugin/k8s_external/external.go (hunk 1 above)
Comment (fturib) [reply plugin/k8s_external/external.go]: See the change in corresponding functions : e.a(...), e.aaa(...). e.srv(...):
- the domain for the A record is usually the one of the `state`, but in case of transfer, it has to the be the one of the `svc.QName()`.
(Unless I am wrong here ... and there is another way).

So either I send state in pointer, either I change signature of those functions, either I duplicate those functions for AXFR use case.

File: plugin/k8s_external/xfr.go (hunk 2 above)
Comment (fturib) [reply plugin/k8s_external/xfr.go]: you mean a new function e.axfr(state) ?

- Maybe that will resolve the point above with pointer of state.

Comment (miekg) [reply plugin/k8s_external/xfr.go]: I think that makes sense, and hopefully also minimizes the changes in k8s

**Comment by fturib** [comment]
Pushed s sqush merge of commits. That includes:
- add setup for "transfer" and update README
- adding AXFR support for k8s_external (with its own translation of the msg.Service)
- change SERIAL of SOA to be timestamp instead of fixed value
- add building the list of All Svc in k8s/external
- fix generation of A record for NS (in k8s/ns)

Needs to do:
- add computation of a proper SERIAL (based on change in the services)
- UT

File: plugin/k8s_external/README.md [hunk 3]
```diff
@@ -51,11 +50,17 @@ this extended syntax.
...
     apex APEX
     ttl TTL
+    transfer to ADDRESS...
 }
 ~~~
 
 * **APEX** is the name (DNS label) to use the apex records, defaults to `dns`.
 * `ttl` allows you to set a custom **TTL** for responses. The default is 5 (seconds).
+* `transfer` enables zone transfers. It may be specified multiples times. `To` signals the direction
+  (only `to` is allow). **ADDRESS** must be denoted in CIDR notation (127.0.0.1/32 etc.) or just as
```
Comment (miekg) [inline plugin/k8s_external/README.md:60]: allowed

File: plugin/k8s_external/msg_to_dns.go [hunk 4]
```diff
@@ -61,10 +61,7 @@ func (e *External) aaaa(services []msg.Service, state request.Request) (records
 	return records
 }
 
-func (e *External) srv(services []msg.Service, state request.Request) (records, extra []dns.RR) {
-	dup := make(map[item]struct{})
-
-	// Looping twice to get the right weight vs priority. This might break because we may drop duplicate SRV records latter on.
+func buildWeights(services []msg.Service) map[int]int {
```
Comment (miekg) [inline plugin/k8s_external/msg_to_dns.go:64]: do we need this? I.e. can there be multiple msg where we need to calculate srv weights?

Also this change seems to be unrelated to the functionality of this pr, so please split it off and make it a global change in all plugins that do this

File: plugin/k8s_external/setup.go [hunk 5]
```diff
@@ -3,6 +3,8 @@ package external
 import (
 	"strconv"
 
+	prs "github.com/coredns/coredns/plugin/pkg/parse"
+
```
Comment (miekg) [inline plugin/k8s_external/setup.go:7]: has extra newline and why are you rewriting the import?

File: plugin/k8s_external/xfr.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+const transferLength = 2000
+
+// Serial implements the Transferer interface.
+func (e *External) Serial(state request.Request) uint32 { return uint32(time.Now().Unix()) }
```
Comment (miekg) [inline plugin/k8s_external/xfr.go:17]: In your PR description you say this is "smart", here it just takes time.Now(). So it's not smart?

File: plugin/k8s_external/xfr.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+	tr.Out(state.W, state.Req, ch)
+	// Defer closing to the client
+	state.W.Hijack()
+	return dns.RcodeSuccess, nil
+}
+
+// transferAllowed checks if incoming request for transferring the zone is allowed according to the ACLs.
+// Note: This is copied from zone.transferAllowed, but should eventually be factored into a common transfer pkg.
+func (e *External) transferAllowed(state request.Request) bool {
```
Comment (miekg) [inline plugin/k8s_external/xfr.go:78]: prime candidate to factor this out (soonish), we have 3 of these (or more) identical functions in our plugins.

File: plugin/kubernetes/ns.go [hunk 8]
```diff
@@ -23,6 +23,8 @@ func (k *Kubernetes) nsAddr() *dns.A {
 	)
 
 	rr := new(dns.A)
+	rr.Hdr.Class = dns.ClassINET
```
Comment (miekg) [inline plugin/kubernetes/ns.go:26]: what does this fix?

File: plugin/kubernetes/external.go [hunk 9]
```diff
@@ -19,6 +19,10 @@ func (k *Kubernetes) External(state request.Request) ([]msg.Service, int) {
 	segs := dns.SplitDomainName(base)
 	last := len(segs) - 1
 	if last < 0 {
+		// this is a request for the zone itself. Return all services that are in exposed namespaces
```
Comment (miekg) [inline plugin/kubernetes/external.go:22]: this is only correct if there is an ixfr/axfr request, this should return the apex of the zone, not all records.

File: plugin/kubernetes/external.go [hunk 10]
```diff
@@ -67,22 +71,51 @@ func (k *Kubernetes) External(state request.Request) ([]msg.Service, int) {
...
 
-				services = append(services, s)
-			}
+func (k *Kubernetes) externalAll(state request.Request) ([]msg.Service, int) {
+
+	serviceList := k.APIConn.ServiceList()
+
+	services := []msg.Service{}
+	zonePath := msg.Path(state.Zone, coredns)
+	rcode := dns.RcodeNameError
```
Comment (miekg) [inline plugin/kubernetes/external.go:88]: this means, when we don't find any services we return nxdomain for the zone it self? That's not correct, should be nodata (at least)

**Comment by miekg** [comment]
> Pushed s sqush merge of commits. That includes:
> 
> * add setup for "transfer" and update README
> * adding AXFR support for k8s_external (with its own translation of the msg.Service)
> * change SERIAL of SOA to be timestamp instead of fixed value
> * add building the list of All Svc in k8s/external
> * fix generation of A record for NS (in k8s/ns)
> 
> Needs to do:
> 
> * add computation of a proper SERIAL (based on change in the services)
> * UT

ah, missed your todo on smart soa serial. that will cause load on the api server though

File: plugin/k8s_external/setup.go (hunk 5 above)
Comment (fturib) [reply plugin/k8s_external/setup.go:7]: you mean : why I name the import "prs" ?
=> there is a confustion with the function parse .. but i can rename that one. ok.

File: plugin/kubernetes/external.go (hunk 9 above)
Comment (fturib) [reply plugin/kubernetes/external.go:22]: I guess that is the responsibility of the caller to reply the correct records depending IXFR/AXFR.
this function `External` here is only to return Services.

File: plugin/kubernetes/external.go (hunk 10 above)
Comment (fturib) [reply plugin/kubernetes/external.go:88]: hum. Anyway this `externallAll` should just return the services.

File: plugin/kubernetes/ns.go (hunk 8 above)
Comment (fturib) [reply plugin/kubernetes/ns.go:26]: that fixes the NS record of the k8s_external.
the function nsAddr() is called by ExternalAddress(..), which provide the RR for NS. And those RR are lacking the Class and Type.

File: plugin/k8s_external/msg_to_dns.go (hunk 4 above)
Comment (fturib) [reply plugin/k8s_external/msg_to_dns.go:64]: my PR is to add AXFR, not to change existing SRV function.
So I keep the same computation ... cannot say really how much that is useful or not.

Comment (fturib) [reply plugin/k8s_external/msg_to_dns.go:64]: but .. sure, I can consider sharing this function in all plugins that have the same computation. Will be in anotehr PR.

**Comment by miekg** [comment]
this is not going anywhere and #2585 (a new plugin) is the better alternative.
